### PR TITLE
[feat] Add trainer builder and trainer API builder

### PR DESCRIPTION
**Pull request by JanLin0817**
* Add trainer builder
* Add trainer api builder
* Add model Hook
* Update trainer test
* Update component test
* Update config yaml
* Update train_cifar10 demo
* Update epoch_based_trainer behavior

### CODE REVIEW

File: pytorch_trainer/trainer/epoch_based_trainer.py [hunk 1]
```diff
@@ -88,6 +88,10 @@ def fit(self, data_loaders, workflow):
                 mode, epochs = flow
                 epoch_trainer = getattr(self, mode)
 
+                # default as -1 means val 1 epoch
+                if epochs == -1:
+                    epochs = 1
```
Comment (mousyball) [inline pytorch_trainer/trainer/epoch_based_trainer.py:93]: Why does epoch need to set to -1? Any reason?

**Review by mousyball — COMMENTED** [review]
GJ

File: pytorch_trainer/trainer/epoch_based_trainer.py (hunk 1 above)
Comment (JanLin0817) [reply pytorch_trainer/trainer/epoch_based_trainer.py:93]: `iter_based_trainer` set to `-1` means run whole train/val set
This make `epoch_based_trainer` has same behavior as `iter_based_trainer`

Comment (mousyball) [reply pytorch_trainer/trainer/epoch_based_trainer.py:93]: OK

**Pull request merged**